sphinx/ext/autodoc/__init__.py

```diff
@@ -1175,22 +1183,70 @@ def format_args(self, **kwargs: Any) -> str:
 
-        # for classes, the relevant signature is the __init__ method's
-        initmeth = self.get_attr(self.object, '__init__', None)
-        # classes without __init__ method, default __init__ or
-        # __init__ written in C?
-        if initmeth is None or \
-                inspect.is_builtin_class_method(self.object, '__init__') or \
-                not(inspect.ismethod(initmeth) or inspect.isfunction(initmeth)):
-            return None
-        try:
-            self.env.app.emit('autodoc-before-process-signature', initmeth, True)
-            sig = inspect.signature(initmeth, bound_method=True)
+        def get_user_defined_function_or_method(obj: Any, attr: str) -> Any:
+            """ Get the `attr` function or method from `obj`, if it is user-defined. """
+            if inspect.is_builtin_class_method(obj, attr):
+                return None
+            attr = self.get_attr(obj, attr, None)
+            if not (inspect.ismethod(attr) or inspect.isfunction(attr)):
+                return None
+            return attr
+
+        sig = None
+
+        # this sequence is copied from inspect._signature_from_callable
+
+        # First, let's see if it has an overloaded __call__ defined
+        # in its metaclass
+        if sig is None:
+            call = get_user_defined_function_or_method(type(self.object), '__call__')
+
+            if call is not None:
+                if "{0.__module__}.{0.__qualname__}".format(call) in _METACLASS_CALL_BLACKLIST:
+                    call = None
+
+            if call is not None:
+                self.env.app.emit('autodoc-before-process-signature', call, True)
+                try:
+                    sig = inspect.signature(call, bound_method=True)
+                except TypeError:
+                    pass
+
+        # Now we check if the 'obj' class has a '__new__' method
+        if sig is None:
+            new = get_user_defined_function_or_method(self.object, '__new__')
+            if new is not None:
+                self.env.app.emit('autodoc-before-process-signature', new, True)
+                try:
+                    sig = inspect.signature(new, bound_method=True)
+                except TypeError:
+                    pass
+
+        # Finally, we should have at least __init__ implemented
+        if sig is None:
+            init = get_user_defined_function_or_method(self.object, '__init__')
+            if init is not None:
+                self.env.app.emit('autodoc-before-process-signature', init, True)
+                try:
+                    sig = inspect.signature(init, bound_method=True)
+                except TypeError:
+                    pass
+
+        # if none of the attributes are user-defined, then it's unlikely that
+        # `autodoc-before-process-signature` will be able to do anything.
+        # Let's just let inspect handle it to ensure we at least get a signature.
+        if sig is None:
+            try:
```

Comment:
```suggestion
            # we don't know the exact location that inspect.signature will find the docstring,
            # so just pass the hook the object itself.
            self.env.app.emit('autodoc-before-process-signature', obj, True)
            try:
```

Reply:
@tk0miya: is this an acceptable compromise?

Reply:
@tk0miya, ping

Reply:
I don't understand what the comment says. And `inspect.signature()` does not refer the docstring, AFAIK. Anyway, we always pass the object itself even if we don't know the exact location.

Reply:
You're right, that wording was bad. Updated.

Reply:
Thanks, Good to me!